### PR TITLE
Component blender re-architecture

### PR DESCRIPTION
**Pull request by jaortiz**
After using the blender I've been finding the class names outputted to be very verbose, fragile and at times hard to understand. I'm proposing a re-architecture of components to create a more traditional css output from the blender. 

Keen to get your thoughts @dominikwilkowski and happy to have a call to discuss. 

### Example Outputs

**Panel**
Current
```
<div class="GEL-panel-look_hero">
    <div class="GEL-panel-header-look_hero">
      <h1 class="GEL-heading-size_9-panel-heading">
        Your hero panel heading
      </h1>
    </div>
    <div class="GEL-body-panel-body">
      Your hero panel content
    </div>
    <div class="GEL-panel-footer">
      Your hero panel footer
    </div>
  </div>
```

Proposed
```
<div class="GEL-panel-hero GEL-panel">
	<div class="GEL-panel-header">
		<h1 class="GEL-panel-heading">Panel title</h1>
	</div>
	<div class="GEL-body-panel-body">
		Lorem ipsum dolor sit amet, consectetur adipisicing elit. Tempora officiis officia omnis aperiam voluptate suscipit, laudantium praesentium quas consequatur placeat, perferendis eligendi saepe in unde sequi dolores excepturi doloremque autem! Lorem ipsum dolor sit amet, consectetur adipisicing elit.
	</div>
	<div class="GEL-PanelFooter">
		I am a footer
	</div>
</div>
```

**Progress Rope**
- Was able to get progress rope working using this new way

Current (output not working correctly)
```
<nav role="navigation" aria-label="In this form" class="GEL-progressRope">
	<ol class="GEL-progressRope-list">
		<li class="GEL-progressRope-step-visited">
			<button type="button" class="GEL-progressRope-stepBtn-visited">
				Step 1
				<span class="GEL-VisuallyHidden">
					, complete
				</span>
			</button>
		</li>
		<li class="GEL-progressRope-step-visited">
			<button type="button" class="GEL-progressRope-stepBtn-visited">
				Step 2
				<span class="GEL-VisuallyHidden">
					, complete
				</span>
			</button>
		</li>
		<li class="GEL-progressRope-step-furthest-visited">
			<button type="button" class="GEL-progressRope-stepBtn-furthest-visited">
				Step 3
				<span class="GEL-VisuallyHidden">
					, complete
				</span>
			</button>
		</li>
		<li class="GEL-progressRope-step-end">
			<button type="button" disabled="" aria-current="step" class="GEL-progressRope-stepBtn-active-end">
				Review and Submit
				<span class="GEL-VisuallyHidden">
					, in progress
				</span>
			</button>
		</li>
	</ol>
</nav>
```

Proposed
```
<nav role="navigation" aria-label="In this form" class="GEL-progressRope">
	<ol class="GEL-progress-rope-list">
		<li index="0" class="GEL-progressRope-step-visited GEL-progressRope-step">
			<button type="button" class="GEL-progressRope-step-btn">
				Step 1
				<span class="GEL-VisuallyHidden">
					, complete
				</span>
			</button>
		</li>
		<li index="1" class="GEL-progressRope-step-visited GEL-progressRope-step">
			<button type="button" class="GEL-progressRope-step-btn">
				Step 2
				<span class="GEL-VisuallyHidden">
					, complete
				</span>
			</button>
		</li>
		<li index="2" class="GEL-progressRope-step-active GEL-progressRope-step">
			<button type="button" aria-current="step" class="GEL-progressRope-step-btn">
				Step 3
				<span class="GEL-VisuallyHidden">
					, in progress
				</span>
			</button>
		</li>
		<li index="5" class="GEL-progressRope-step-end GEL-progressRope-step">
			<button type="button" disabled="" class="GEL-progressRope-step-btn">
				Review and Submit
				<span class="GEL-VisuallyHidden">
					, not started
				</span>
			</button>
		</li>
	</ol>
</nav>
```

### CODE REVIEW

**Comment by jaortiz** [comment]
@dominikwilkowski I've done some rework based on your comments. Is there a way in the blender we can specify a base version of a component (default props) and use that to perform a diff on any further variations of a component removing duplicate styles and leaving the different styles as that props class styling? The way I've rearchitected this is based on the assumption that we can do this.

File: components/button/src/Button.js [hunk 1]
```diff
@@ -127,7 +129,7 @@ Button.propTypes = {
 	/**
 	 * Button look
 	 */
-	look: PropTypes.oneOf(['primary', 'hero', 'faint', 'link']),
+	look: PropTypes.oneOf(['primary', 'hero', 'faint', 'link', null]),
```
Comment (dominikwilkowski) [inline components/button/src/Button.js:132]: with [default prop being `hero`](https://github.com/WestpacGEL/GEL/blob/2b1cf0075fdd665e50da48ced5603c2075aa93c3/components/button/src/Button.js#L236) does this make sense?

File: components/button/src/overrides/content.js [hunk 2]
```diff
@@ -6,7 +6,7 @@ import { Fragment } from 'react';
 const Content = ({ state, children }) => <Fragment>{children}</Fragment>;
 
 const contentStyles = (_, { hasChildren }) => ({
```
Comment (dominikwilkowski) [inline components/button/src/overrides/content.js:8]:
```suggestion
const contentStyles = () => ({
```
no need for `hasChildren`

File: components/button/src/overrides/button.js [hunk 3]
```diff
@@ -1,11 +1,76 @@
...
+// Component
+// ==============================
+
 const Button = forwardRef(({ state: { tag: Tag }, ...rest }, ref) => <Tag ref={ref} {...rest} />);
 
-const buttonStyles = (_, { look, size, soft, block, justify, disabled }) => {
+// ==============================
+// Styles
+// ==============================
+const blenderStyles = (_, { look, size, soft, block, justify, disabled }) => {
```
Comment (jaortiz) [inline components/button/src/overrides/button.js:15]: @dominikwilkowski Here's my attempt of diff'ing modifier styles with the base styles from default props. Is working correctly for me with the blender. Thoughts?

File: components/button/src/overrides/button.js [hunk 4]
```diff
@@ -1,11 +1,76 @@
...
+	if (!modifiers.length) return baseStyles;
+
+	const modifiedStyles = buttonStyles(_, props);
+
+	// find all different
+	// we need to go deeper...for nested styles?
+	const diff = Object.keys(modifiedStyles).filter((m) => baseStyles[m] !== modifiedStyles[m]);
+
+	let modifier,
+		label = baseStyles.label;
```
Comment (dominikwilkowski) [inline components/button/src/overrides/button.js:40]:
```suggestion
	let modifier;
	let label = baseStyles.label;
```

File: components/button/src/overrides/button.js [hunk 5]
```diff
@@ -1,11 +1,76 @@
...
+	// find the prop that is different from default
+	let modifiers = Object.keys(defaultProps).filter((m) => defaultProps[m] !== props[m]);
+
+	if (!modifiers.length) return baseStyles;
+
+	const modifiedStyles = buttonStyles(_, props);
+
+	// find all different
+	// we need to go deeper...for nested styles?
+	const diff = Object.keys(modifiedStyles).filter((m) => baseStyles[m] !== modifiedStyles[m]);
```
Comment (dominikwilkowski) [inline components/button/src/overrides/button.js:37]: Need to make this work with nested styles

**Comment by jaortiz** [comment]
@dominikwilkowski I've re-done badge, button, panel and progress-rope based on the feedback from you before. Only issue is the use of the `.GEL-` prefix hardcoded in some places. Is it possible for the blender to prefix any nested classes as well? Keen to hear your thoughts.

**Pull request merged**